app/src/main/java/org/oppia/app/utility/Logger.kt

```diff
@@ -0,0 +1,115 @@
+
+  private val blockingScope = CoroutineScope(blockingDispatcher)
+
+  private val ENABLE_CONSOLE_LOG = true
+  private val ENABLE_FILE_LOG = true
+  private val GLOBAL_LOG_LEVEL = LogLevel.VERBOSE
+  private val LOG_DIRECTORY = File(context.filesDir, "oppia_app.log")
+
+  private enum class LogLevel private constructor(val logLevel: Int) {
+    VERBOSE(Log.VERBOSE),
+    DEBUG(Log.DEBUG),
+    INFO(Log.INFO),
+    WARNING(Log.WARN),
+    ERROR(Log.ERROR),
+    ASSERT(Log.ASSERT)
+  }
+
+  /** Logs a verbose message with the specified tag.*/
+  fun v(tag: String, msg: String) {
+    writeLog(LogLevel.VERBOSE, tag, msg)
+  }
+
+  /** Logs a verbose message with the specified tag, message and exception.*/
+  fun v(tag: String, msg: String, tr: Throwable) {
+    writeError(LogLevel.VERBOSE, tag, msg, tr)
+  }
+
+  /** Logs a debug message with the specified tag*/
+  fun d(tag: String, msg: String) {
+    writeLog(LogLevel.DEBUG, tag, msg)
+  }
+
+  /** Logs a debug message with the specified tag, message and exception.*/
+  fun d(tag: String, msg: String, tr: Throwable) {
+    writeError(LogLevel.DEBUG, tag, msg, tr)
+  }
+
+  /** Logs a info message with the specified tag.*/
+  fun i(tag: String, msg: String) {
+    writeLog(LogLevel.INFO, tag, msg)
+  }
+
+  /** Logs a info message with the specified tag, message and exception.*/
+  fun i(tag: String, msg: String, tr: Throwable) {
+    writeError(LogLevel.INFO, tag, msg, tr)
+  }
+
+  /** Logs a warn message with the specified tag.*/
+  fun w(tag: String, msg: String) {
+    writeLog(LogLevel.WARNING, tag, msg)
+  }
+
+  /** Logs a warn message with the specified tag, message and exception.*/
+  fun w(tag: String, msg: String, tr: Throwable) {
+    writeError(LogLevel.WARNING, tag, msg, tr)
+  }
+
+  /** Logs a error message with the specified tag.*/
+  fun e(tag: String, msg: String) {
+    writeLog(LogLevel.ERROR, tag, msg)
+  }
+
+  /** Logs a error message with the specified tag, message and exception.*/
+  fun e(tag: String, msg: String, tr: Throwable) {
+    writeError(LogLevel.ERROR, tag, msg, tr)
+  }
+
+  private fun isLogEnable(logLevel: LogLevel): Boolean {
+    return GLOBAL_LOG_LEVEL.logLevel < logLevel.logLevel
+  }
+
+  private fun writeLog(logLevel: LogLevel, tag: String, log: String) {
+    writeInternal(logLevel, tag, log)
+  }
+
+  private fun writeError(logLevel: LogLevel, tag: String, log: String, tr: Throwable) {
+    writeInternal(logLevel, tag, "$log\n${Log.getStackTraceString(tr)}")
+  }
+
+  private fun writeInternal(logLevel: LogLevel, tag: String, fullLog: String) {
+    if (isLogEnable(logLevel) && ENABLE_CONSOLE_LOG) {
+      Log.println(logLevel.logLevel, tag, fullLog)
+    }
+    if (isLogEnable(logLevel) && ENABLE_FILE_LOG) {
+      val msg = "${Calendar.getInstance().time}\t${logLevel.name}/$tag: $fullLog"
+
+      // To mange background threads
```

Comment:
A more descriptive comment might be:

```kotlin
// To ensure that saving messages don't block the main thread, and are saved in order.
```

The current comment as-is doesn't provide much context into what specifically is being managed here.

Reply:
done